### DISCUSSION TITLE
Trivy CloudFormation scanner skips YAML files with CloudFormation-specific functions

### DISCUSSION THREAD

**Discussion by GeorgeDavis-Ibexlabs** [aquasecurity/trivy#7326] [Bugs]
### Description

As identified in #3418, `trivy` skips CloudFormation files that have YAML parsing errors. This includes any file that have template section [Conditions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/conditions-section-structure.html) or other CloudFormation-specific [Intrinsic functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) like `!Ref` will be skipped when performing a scan.

### Desired Behavior

`trivy` scans all YAML CloudFormation templates with the understanding of CloudFormation-specific functions. 

### Actual Behavior

`trivy` skips configuration files that have YAML parsing errors

### Reproduction Steps

```bash
1. Write a CloudFormation template with the `!Ref` function used in it
2. Scan the configuration file with `trivy config`
3. `trivy` skips the file and prints `INFO    Detected config files   num=0`
```

### Target

Git Repository

### Scanner

Misconfiguration

### Output Format

SARIF

### Mode

Standalone

### Debug Output

```bash
trivy repo -c .github/config/trivy-sarif.yaml --debug .
2024/08/09 19:16:34 INFO Loaded file_path=/Users/***/***/.github/config/trivy-sarif.yaml
2024-08-09T19:16:34-04:00       DEBUG   Cache dir       dir="/Users/***/Library/Caches/trivy"
2024-08-09T19:16:34-04:00       DEBUG   Parsed severities       severities=[HIGH CRITICAL]
2024-08-09T19:16:34-04:00       INFO    [misconfig] Misconfiguration scanning is enabled
2024-08-09T19:16:34-04:00       DEBUG   [misconfig] Policies successfully loaded from disk
2024-08-09T19:16:34-04:00       DEBUG   Enabling misconfiguration scanners      scanners=[azure-arm cloudformation dockerfile helm kubernetes terraform terraformplan-json terraformplan-snapshot]
2024-08-09T19:16:34-04:00       DEBUG   Initializing scan cache...      type="memory"
2024-08-09T19:16:34-04:00       DEBUG   Scanning files for misconfigurations... scanner="CloudFormation"
2024-08-09T19:16:34-04:00       DEBUG   OS is not detected.
2024-08-09T19:16:34-04:00       INFO    Detected config files   num=0
2024-08-09T19:16:34-04:00       DEBUG   [vex] VEX filtering is disabled
```

### trivy-sarif.yaml
```
timeout: 10m
format: sarif
output: trivy-results.sarif
severity:
    - HIGH
    - CRITICAL
scan:
    scanners:
        - vuln
        - secret
        - misconfig
    ignore-unfixed: false
    hide-progress: false
    skip-dirs:
        - ansible
```

### Operating System

macOS Sonoma 14.6

### Version

```bash
Version: 0.54.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-11-22 18:12:00.614281014 +0000 UTC
  NextUpdate: 2023-11-23 00:12:00.614280854 +0000 UTC
  DownloadedAt: 2023-11-22 19:49:54.681965 +0000 UTC
Check Bundle:
  Digest: sha256:ef2d9ad4fce0f933b20a662004d7e55bf200987c180e7f2cd531af631f408bb3
  DownloadedAt: 2024-08-09 18:37:56.333557 +0000 UTC
```

### Checklist

- [X] Run `trivy clean --all`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by nikpivkin** [comment]
Hi @GeorgeDavis-Ibexlabs ! 

Can you give an example of a configuration file that is not scanned by Trivy? Right now Trivy supports CloudFormation syntax (see [tests](https://github.com/aquasecurity/trivy/blob/main/pkg/iac/scanners/cloudformation/parser/parser_test.go#L183)).

**Comment by GeorgeDavis-Ibexlabs** [comment]
Hello @nikpivkin Thank you for taking time to look into this issue. Here is a simple CloudFormation template in YAML that ends up returning the following output (output is truncated to show the detection numbers).

```bash
2024-08-12T09:00:46-04:00       INFO    Number of language-specific files       num=0
2024-08-12T09:00:46-04:00       INFO    Detected config files   num=0
2024-08-12T09:00:46-04:00       DEBUG   [vex] VEX filtering is disabled
```


Simple CloudFormation template in YAML -
```yaml
---
AWSTemplateFormatVersion: '2010-09-09'
Description: This is a simple YAML template for trivy discussion (https://github.com/aquasecurity/trivy/discussions/7326)

Parameters:
  SimpleParameter:
    Description: This is a simple string parameter.
    Type: String
  
Resources:
  SimpleBucket:
    Type: AWS::S3::Bucket
    DeletionPolicy: Delete
    Properties:
      BucketName: !Ref SimpleParameter

Outputs:
  SimpleParameter:
    Description: This is the simple parameter value passed to the CloudFormation template
    Value: !Ref SimpleParameter
  
  SimpleBucket:
    Description: This is the S3 Bucket name created with the Simple Parameter value
    Value: !Ref SimpleBucket
```

**Reply by nikpivkin** [reply]
Thanks. 
The template uses variables, but you have not passed them, so the misconfigurations associated with the bucket are not detected. You [can pass](https://aquasecurity.github.io/trivy/v0.54/docs/coverage/iac/cloudformation/#value-overrides) parameters using the `--cf-params` flag.  I think we should add a log about the absence of passed parameters.

Regarding the message that 0 configuration files are detected. The message is misspelled and this [issue](https://github.com/aquasecurity/trivy/issues/6138) will be fixed soon.

**Reply by GeorgeDavis-Ibexlabs** [reply]
Thank you @nikpivkin 

I wrote a quick script to generate the `cf-params` file and can be used with any scan. The script infers a suitable parameter value to assume so we can proceed with scanning the repository with `trivy`. You can find this side project [here](https://github.com/GeorgeDavis-Ibexlabs/generate-cf-params-file).

I am yet to validate if all files are being scanned with the newly injected parameters. Will keep this thread posted with my test outputs.
